iroha/src/torii/utils.rs

```diff
@@ -0,0 +1,96 @@
+    fn into_response(self) -> Response {
+        Response::new(vec![].into())
+    }
+}
+
+/// Structure for response in scale codec in body
+pub struct Scale<T>(pub T);
+
+impl<T: Encode + Send> Reply for Scale<T> {
+    fn into_response(self) -> Response {
+        Response::new(self.0.encode().into())
+    }
+}
+
+/// Adds state to filter
+pub fn add_state<State: Send + Clone>(
+    state: State,
+) -> impl Filter<Extract = (State,), Error = Rejection> + Clone + Send {
+    warp::any().and_then(move || {
+        let state = state.clone();
+        async move { Ok::<_, Rejection>(state) }
+    })
+}
+
+pub mod body {
+    use super::*;
+
+    //pub fn scale<T: Decode>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy {
+    //    use warp::hyper::body::Bytes;
+    //
+    //    #[derive(Debug)]
+    //    struct ScaleErr(parity_scale_codec::Error);
+    //
+    //    impl warp::reject::Reject for ScaleErr {}
+    //
+    //    warp::body::bytes().and_then(|body: Bytes| async move {
+    //        T::decode(&mut body.as_ref()).map_err(|err| warp::reject::custom(ScaleErr(err)))
+    //    })
+    //}
+
+    /// Decode query request
+    pub fn query() -> impl Filter<Extract = (VerifiedQueryRequest,), Error = Rejection> + Copy {
+        warp::body::bytes()
+            .and_then(|body: Bytes| async move { (&body).try_into().map_err(warp::reject::custom) })
+    }
+
+    /// Decode body as versioned scale codec
+    pub fn versioned<T: DecodeVersioned>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy
+    {
+        warp::body::bytes().and_then(|body: Bytes| async move {
+            DecodeVersioned::decode_versioned(body.as_ref()).map_err(warp::reject::custom)
+        })
+    }
+}
+
+macro_rules! impl_custom_and_then {
+    ( $name:ident ( $($arg_name:ident : $arg_gen:ident),* $(,)? ) ) => {
+        /// Maps filter to handler with `n` arguments (`n` is suffix of function)
+        pub fn $name<O, E, F, Fut, Fil, $($arg_gen,)*>(f: F, router: Fil) -> impl Filter<Extract = (O,), Error = Rejection> + Clone
+        where
+            Fil: Filter<Extract = ($($arg_gen,)*), Error = Rejection> + Clone,
+            F: Fn($($arg_gen,)*) -> Fut + Copy + Send + Sync + 'static,
+            Fut: std::future::Future<Output = Result<O, E>> + Send,
+            E: Reject,
+            $($arg_gen: Send,)*
+        {
+            router.and_then(move |$($arg_name,)*|
+                async move {
+                    f($($arg_name,)*).await.map_err(warp::reject::custom)
+                }
+            )
+        }
+    }
+}
+
+//impl_custom_and_then!(endpoint1 (a: A));
+impl_custom_and_then!(endpoint2(a: A, b: B));
+impl_custom_and_then!(endpoint3(a: A, b: B, c: C));
+//impl_custom_and_then!(endpoint4 (a: A, b: B, c: C, d: D));
+
+macro_rules! warp {
+    ( and -> $start:expr, $($filter:expr),* $(,)? ) => { $start $(.and($filter))* };
+    ( or  -> $start:expr, $($filter:expr),* $(,)? ) => { $start $(.or($filter))*  };
+}
```

Comment:
Considering `impl_custom_and_then`:
Could you do the same kind of thing (rust lacks currying) as elegent as this functions do? I just didnt find other nice way to do it.

As for `warp` macro it just removes boilerplate (sequence of `.and()`-s and `.or()`-s)

Reply:
For `warp macro you still write same `and` and `or`. But it adds complexity that you have to go and check the macro.

For the first point can you show how the code would look like without this macro? So we can understand what complexity we are trying to eliminate here.

Reply:
A quote from Rust book:
```
The downside to implementing a macro instead of a function is that macro definitions are more complex than function definitions because you’re writing Rust code that writes Rust code. Due to this indirection, macro definitions are generally more difficult to read, understand, and maintain than function definitions.
```

The thing is when macros do not significantly simplify our life - it is better not to write them. Or it makes code rather hard to understand.

Reply:
We don't have many endpoints, so even if we will write this* for each of them it still will be better than having this big macro

* - 
```rust
router.and_then(move |$($arg_name,)*|
                async move {
                    f($($arg_name,)*).await.map_err(warp::reject::custom)
                }
            )
```

Reply:
> For `warp macro you still write same `and`and`or`. But it adds complexity that you have to go and check the macro.
I dont think so. You dont look at the meaning of every macro which you dont know. This one
Generally i dont think you should though. The main point that it looks way nicer than combinator repetition.
Also macro is 4 lines of code, so it is very understandable.

> For the first point can you show how the code would look like without this macro? So we can understand what complexity we are trying to eliminate here.
This is done everytime final endpoint is called:
```rust
            router.and_then(move |a, b, c, d|
                async move {
                    f(a, b, c, d).await.map_err(warp::reject::custom)
                }
            )
```

Reply:
It already has complexity in this part `( and -> $start:expr, $($filter:expr),* $(,)? ) => { $start $(.and($filter))* };`. To my mind it doesn't look nicer - it just confuses - as it looks different from the normal Rust code.

Reply:
We already have exactly this methods, that do what we need :)

Reply:
> It already has complexity in this part `( and -> $start:expr, $($filter:expr),* $(,)? ) => { $start $(.and($filter))* };`. To my mind it doesn't look nicer - it just confuses - as it looks different from the normal Rust code.

I just dont see a complexity where you point. It is readable if you've read rust book.